schematic/__init__.py

```diff
@@ -46,6 +52,107 @@
+        }
+        self.export = export
+
+    def _redact_string(self, value: str) -> str:
+        """remove sensitive data from a string
+
+        Args:
+            value (str): a string that may contain sensitive data
+
+        Returns:
+            str: remove sensitive data from string
+        """
+        redacted = value
+        for pattern_name, pattern in self._compiled_patterns.items():
+            redacted = pattern.sub(f"[REDACTED_{pattern_name.upper()}]", redacted)
+        return redacted
+
+    def _redacted_sensitive_data_in_exception(
+        self, exception_attributes: Dict[str, str]
+    ) -> Dict[str, str]:
+        """remove sensitive data in exception
+
+        Args:
+            exception_attributes (dict):a dictionary of exception attributes
+
+        Returns:
+            dict: a dictionary of exception attributes with sensitive data redacted
+        """
+        redacted_exception_attributes = {}
+        for key, value in exception_attributes.items():
+            # remove sensitive information from exception message and stacktrace
+            if key == "exception.message" or key == "exception.stacktrace":
+                redacted_exception_attributes[key] = self._redact_string(value)
+            else:
+                redacted_exception_attributes[key] = value
+        return redacted_exception_attributes
+
+    def _create_redacted_span(self, span: ReadableSpan) -> ReadableSpan:
+        redacted_events = []
+        for event in span.events:
+            attributes = event.attributes
+            # remove sensitive data from exception
+            redacted_event_attributes = self._redacted_sensitive_data_in_exception(
+                attributes
+            )
+            redacted_event = Event(
+                name=self._redact_string(event.name),
+                attributes=redacted_event_attributes,
+                timestamp=event.timestamp,
+            )
+            redacted_events.append(redacted_event)
+            # Create new span with redacted data
+        redacted_span = ReadableSpan(
+            name=span.name,
+            context=span.get_span_context(),
+            parent=span.parent,
+            resource=span.resource,
+            attributes=span.attributes,
+            events=redacted_events,
+            links=span.links,
+            kind=span.kind,
+            status=span.status,
+            start_time=span.start_time,
+            end_time=span.end_time,
+            instrumentation_info=span.instrumentation_info,
+        )
+        return redacted_span
+
+    def on_end(self, span: ReadableSpan) -> None:
+        """Remove sensitive information in spans by creating new ones with redacted data."""
+        # maybe the goal should be to filter all spans?
+        if span.status.status_code == trace.StatusCode.ERROR:
+            if span.events:
+                redacted_span = self._create_redacted_span(span)
+                self.export.export([redacted_span])
```

Comment:
```suggestion
        if span.status.status_code == trace.StatusCode.ERROR:
            if span.events:
                redacted_span = self._create_redacted_span(span)
                self.export.export([redacted_span])
                return
        self.export.export([span])
```

Reply:
Is this needed? I'm not sure tbh

Reply:
I don't think this is going to work with how we have this set up because of:
https://github.com/open-telemetry/opentelemetry-python/blob/a7fe4f8bac7fa36291c6acf86982bbb356e3ae6d/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L173-L175

**Specifically:**

1. We have a `BatchSpanProcessor` that was already responsible for queuing up and exporting the spans to SigNoz
2. Based on the code at the link above the span is sent to each processor, in this case the data is likely being sent twice. Once in the `BatchSpanProcessor` that is still being called, and once here.

Reply:
Here is a hack how we can get around this issue:

1. Instead of using a processor to handle this logic, as we can see, won't work we could monkey patch the `_readable_span` function call: https://github.com/open-telemetry/opentelemetry-python/blob/a7fe4f8bac7fa36291c6acf86982bbb356e3ae6d/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L906-L921
2. In the monkey patch we: 1) Start by calling our sensitive data redaction process to strip data out of the span, 2) Return a call to the original function.

By monkey patching this, it would allow us to essentially "slip" in logic before a read-only span has been created.

Let me know what questions you have.